Created flake for project

inkpot-monkey opened the pull request:
Simple flake that create a Nix package for the project, creates apps to run the browser versions of the project. A dev shell for for any future dev and some checks that utilizes the general test file

inkpot-monkey committed:
Created flake for project

DieracDelta commented:
LGTM. I built and run the apps and checks. Do you know what the difference is between wallet and walletclear?

Only slightly issue I encountered was if the XDG browser variable is set to something that doesn't exist, the XDG script will barf. I think that's fine, since that's a runtime event. Could you add to the readme that `BROWSER` needs to be set to an installed application in order to nix run (/include brief instructions on how to run. E.g. `BROWSER=$INSTALLED_BROWSER nix run ...` ).

inkpot-monkey committed:
Use current directory as source

inkpot-monkey committed:
Project has no versioning

inkpot-monkey committed:
Made not about usage of wallet-clear

inkpot-monkey commented:
The only difference between the 2 wallets is that the default `wallet` serves the javascript minified while `wallet-clear` doesnt so its readable in the source.

Ive made some notes in the flake itself about the XDG browser variable do you think that this should go into the main Readme? Feels like its more discoverable in the flake as the Readme is already quite chunky...

What would be really good is if we could add documentation to the flake that could be queried through `nix flake show --verbose`
Documentation for every output of the flake could then be combined into a mini man page for the flake.

inkpot-monkey commented:
Another thought that I had was creating a  .desktop file for the project so that the end user need never enter the realm of the shell. What do you reckon? Worth it?

inkpot-monkey committed:
Add documentation about the flake apps

DieracDelta commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+        # Additional tests, if applicable.
+        test = with nixpkgsFor.${system};
+          stdenv.mkDerivation {
+            name = "bitcoin-transactions-test-${version}";
```
I think this variable (`version`) is missing

inkpot-monkey replied:
Ah yeah I removed this as there is no clear versioning on the original project (forget to remove it in this test section) 

What are your thoughts on versioning for this package? Is it needed or do we just go for the last time the repo was last modified?

DieracDelta replied:
Yeah, doesn't look like there is a release. Last time the repo was modified makes sense on my end.

DieracDelta commented on flake.lock:
```diff
@@ -0,0 +1,39 @@
+{
+  "nodes": {
+    "bitcoin-transactions-src": {
```
The narhash is out of date for me. It's capitalizing on annoying bug with `path` where the narhash needs to be updated every  time the repo changes. You'd need to run `nix flake lock --update-input bitcoin-transactions-src` to correct.

DieracDelta commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+  inputs.nixpkgs.url = "nixpkgs/nixos-unstable"; # needed for nodePackage.terser
+
+  # Upstream source tree(s).
+  inputs.bitcoin-transactions-src = {
```
I think self could be used.

DieracDelta replied:
EDIT or `./.`

DieracDelta commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+          in stdenv.mkDerivation {
+            name = "bitcoin-transactions";
+
+            src = bitcoin-transactions-src;
```
`src = lib.cleanSource ./.`

DieracDelta commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+            }/bin/bitcoin-transactions";
+        };
+        # This shows the Javascript unminified and readable
+        wallet-clear = let pkgs = nixpkgsFor.${system};
```
This might be overkill, but since the program here is the same for both, you could make a function `genProgram` which takes in a parameter specifying `walletclear` or `wallet` and generates the program based on that.

inkpot-monkey replied:
What is the advantage to writing a nix function to generate a program as opposed to adding flags to the `ShellScriptBin`? 

For instance 
```
          program = "${
              pkgs.writeShellScriptBin "bitcoin-transactions" ''
                # Clear shows the Javascript unminified and readable
                if [ "$1" == "clear" ] || [ "$1" == "c" ]; then
                  ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/html/walletclear.html
                else
                  ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/html/wallet.html
                fi
              ''
            }/bin/bitcoin-transactions";
```

Which could then be called with `nix run .#wallet -- c` etc.

DieracDelta replied:
Ooh, yeah. I like this more. +1. I think only the path would need to be set (since the invocation is the same for both).
```bash
          program = "${
              pkgs.writeShellScriptBin "bitcoin-transactions" ''
                # Clear shows the Javascript unminified and readable
                if [ "$1" == "clear" ] || [ "$1" == "c" ]; then
                  PAGE=html/walletclear.html
                else
                  PAGE=html/wallet.html
                fi
                ${pkgs.xdg-utils}/bin/xdg-open ${pkgs.bitcoin-transactions}/$PAGE
              ''
            }/bin/bitcoin-transactions";
```

inkpot-monkey replied:
Nice catch ill make that change now :+1:

DieracDelta commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+
+              # Run all general tests apart from the one that checks the network connection
+              while read line; do
+                if [[ $line == node* ]] && [[ $line != *suprnova* ]]; then
```
Nitpick: quotes around `$line` is good practice. Otherwise here's an edge case (perhaps not in this scenario b/c you're in a for loop) because  `line` is an empty string. This results in something like `if [[ == node* ]] ...`. This will fail with a syntax error, instead of returning false, which is the intended behavior.

DieracDelta commented:
+1 for an output that creates a .desktop file. I think that's pretty elegant.

As for where the docs go, in the flake should be fine. Your comments in there LGTM.

inkpot-monkey committed:
Fixed narhash being out of date

inkpot-monkey committed:
Feedback changes

1. src now equals lib.cleanSource ./.
2. supportedSystems expanded to any system that can run a browser
3. combined apps into single script with flags
4. created nixosModule which creates a .desktop file
5. removed unused version var
6. enclose bash vars in ""

inkpot-monkey committed:
extract html into var

inkpot-monkey committed:
Version now equals last git commit

inkpot-monkey committed:
Src = self

DieracDelta commented:
LGTM. Nice work!

aaronjanse commented on flake.nix:
```diff
@@ -0,0 +1,154 @@
+              # Run all general tests apart from the one that checks the network connection
+              while read line; do
+                if [[ "$line" == node* ]] && [[ "$line" != *suprnova* ]]; then
+                  $line
```
Could we use `sed -i` here to comment-out the lines we don't want to run? I'm concerned about future version of the package having a script with something like a multi-line `if` statement.

inkpot-monkey replied:
I dont think this is much of an issue as the project has stopped development but I agree that this could get messy quickly and is not the best way to do things :)

How does this look as an alternative?
```
            buildPhase = ''
              echo 'running some integration tests'
              ln -s ${bitcoin-transactions}/* .

              cp tests/general.js integration-tests.js

              # Run all tests apart from the one that checks the network connection
              sed -i '
                  /^\(node\)/!d
                  /suprnova/d' integration-tests.js

              while read line; do
                  $line
              done < integration-tests.js
            '';
```

aaronjanse replied:
Would `bash integration-tests.js` work here, instead of the `while` loop?

aaronjanse commented on flake.nix:
```diff
@@ -0,0 +1,154 @@
+        in {
+          type = "app";
+          program = "${
+              pkgs.writeShellScriptBin "bitcoin-transactions" ''
```
Perhaps we could use [`writeShellScript`](https://github.com/NixOS/nixpkgs/blob/3e0ce8c5d478d06b37a4faa7a4cc8642c6bb97de/pkgs/build-support/trivial-builders.nix#L203-L226) here?

aaronjanse commented on flake.nix:
```diff
@@ -0,0 +1,154 @@
+          defaultPackage = self.defaultPackage.${pkgs.system};
+        in {
+          environment.systemPackages = [
+            (pkgs.writeTextFile {
```
Using [`makeDesktopItem`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/make-desktopitem/default.nix) might make this easier. There's some [example usage](https://github.com/NixOS/nixpkgs/blob/3b6c3bee9174dfe56fd0e586449457467abe7116/pkgs/games/maelstrom/default.nix#L23-L30) in nixpkgs's maelstorm package

inkpot-monkey replied:
ooo didnt know about this!

Side question do you reckon this would be best added to the main package and remove the nixosModule?

aaronjanse commented on flake.nix:
```diff
@@ -0,0 +1,154 @@
+      devShell = forAllSystems (system:
+        let
+          pkgs = nixpkgsFor.${system};
+          inherit (pkgs) mkShell nodePackages;
```
```suggestion
          inherit (nixpkgsFor.${system}) mkShell nodePackages;
```

aaronjanse reviewed (changes requested):
Thank you @thomasSDK for writing this! I left a few comments that might help make the expression more concise

inkpot-monkey committed:
Convert ShellScriptBin to ShellScript

inkpot-monkey committed:
simplify inherit

Co-authored-by: Aaron Janse <aaron@ajanse.me>

inkpot-monkey committed:
Refactor tests

inkpot-monkey committed:
make app defaultApp

inkpot-monkey committed:
Make use of `makeDesktopItem` function

inkpot-monkey committed:
Tidy

inkpot-monkey commented:
Made those suggested changes @aaronjanse and converted the only app of the flake into the defaultApp to simplify things further

aaronjanse reviewed (approved):
Everything here LGTM. Thank you @thomasSDK!
